### Covid Cases By State

https://www.cdc.gov/covid-data-tracker/#cases

How can we find states that have similar numbers of COVID cases (total or per capita)? 
For example, let's write a SQL query to find all states that have similar infection numbers. 

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/US_MAP_DATA.csv')

In [3]:
df

abbr  fips                    jurisdiction  Total Cases  Total Death  \
0    AK     2                          Alaska         2729           22   
1    AL     1                         Alabama        82530         1493   
2    AR     5                        Arkansas        40181          428   
3    AS    60                  American Samoa            0            0   
4    AZ     4                         Arizona       165934         3408   
5    CA     6                      California       466550         8518   
6    CO     8                        Colorado        45314         1807   
7    CT     9                     Connecticut        49077         4423   
8    DC    11            District of Columbia        11945          583   
9    DE    10                        Delaware        14602          581   
10   FL    12                         Florida       436867         6117   
11  FSM    64  Federated States of Micronesia            0            0   
12   GA    13                         Georgia       175052         3563   
13   GU    66                            Guam          351            5   
14   HI    15                          Hawaii         1647           26   
15   IA    19                            Iowa        42782          839   
16   ID    16                           Idaho        19222          160   
17   IL    17                        Illinois       174973         7638   
18   IN    18                         Indiana        63678         2924   
19   KS    20                          Kansas        26172          335   
20   KY    21                        Kentucky        28126          719   
21   LA    22                       Louisiana       111038         3812   
22   MA    25                   Massachusetts       116182         8551   
23   MD    24                        Maryland        86285         3478   
24   ME    23                           Maine         3866          121   
25   MI    26                        Michigan        87958         6421   
26   MN    27                       Minnesota        52947         1629   
27   MO    29                        Missouri        44823         1213   
28   MP    69        Northern Mariana Islands           40            2   
29   MS    28                     Mississippi        55804         1563   
..  ...   ...                             ...          ...          ...   
31   NC    37                  North Carolina       116087         1820   
32   ND    38                    North Dakota         6227          102   
33   NE    31                        Nebraska        25157          321   
34   NH    33                   New Hampshire         6500          409   
35   NJ    34                      New Jersey       180295        15825   
36   NM    35                      New Mexico        19791          626   
37   NV    32                          Nevada        45151          785   
38   NY    36                        New York       188629         8826   
39  NYC    57                   New York City       225850        23512   
40   OH    39                            Ohio        86497         3382   
41   OK    40                        Oklahoma        34990          503   
42   OR    41                          Oregon        17416          303   
43   PA    42                    Pennsylvania       110218         7162   
44   PR    72                     Puerto Rico        16061          211   
45   PW    70                           Palau            0            0   
46   RI    44                    Rhode Island        18725         1005   
47  RMI    68    Republic of Marshall Islands            0            0   
48   SC    45                  South Carolina        84109         1565   
49   SD    46                    South Dakota         8492          123   
50   TN    47                       Tennessee        99044          999   
51   TX    48                           Texas       394265         5877   
52   UT    49               

In [4]:
from pandasql import sqldf
pysqldf = lambda q: sqldf(q, globals())

You may notice that "Total Cases" has a space in it. Spaces aren't allowed in SQL when referring to table names. A quick fix here is to put back ticks aroudn it, `Total Cases`. Note that these are ticks, not quotes! (you'll generally find them in the upper left corner of your keyboard). 

In [5]:
pysqldf("""
SELECT 
    abbr, `Total Cases`
FROM 
    df
""")

abbr  Total Cases
0    AK         2729
1    AL        82530
2    AR        40181
3    AS            0
4    AZ       165934
5    CA       466550
6    CO        45314
7    CT        49077
8    DC        11945
9    DE        14602
10   FL       436867
11  FSM            0
12   GA       175052
13   GU          351
14   HI         1647
15   IA        42782
16   ID        19222
17   IL       174973
18   IN        63678
19   KS        26172
20   KY        28126
21   LA       111038
22   MA       116182
23   MD        86285
24   ME         3866
25   MI        87958
26   MN        52947
27   MO        44823
28   MP           40
29   MS        55804
..  ...          ...
31   NC       116087
32   ND         6227
33   NE        25157
34   NH         6500
35   NJ       180295
36   NM        19791
37   NV        45151
38   NY       188629
39  NYC       225850
40   OH        86497
41   OK        34990
42   OR        17416
43   PA       110218
44   PR        16061
45   PW            0
46   RI        18725
47  RMI            0
48   SC        84109
49   SD         8492
50   TN        99044
51   TX       394265
52   UT        39337
53   VA        87993
54   VI          385
55   VT         1405
56   WA        54205
57   WI        54064
58   WV         6173
59   WY         2589
60   US      4339997

[61 rows x 2 columns]

If you'd rather not deal with this repeatedly, you can rename the column in SQL or pandas - we'll use pandas here.

In [6]:
df.rename(columns={'Total Cases': 'Total_Cases'}, inplace=True)

Let's start by finding states that have total cases within 10000 of each other. 

In [7]:
within_10000 = pysqldf("""
SELECT
    a.abbr as abbr_a, 
    b.abbr as abbr_b, 
    a.Total_Cases as total_a,
    b.Total_Cases as total_b
FROM   
    df a
JOIN 
    df b 
ON
    ABS(total_a - total_b) < 10000
AND
    abbr_a <> abbr_b
AND
    abbr_a > abbr_b
""")

In [8]:
within_10000

abbr_a abbr_b  total_a  total_b
0       AS     AK        0     2729
1       CO     AR    45314    40181
2       CT     AR    49077    40181
3       CT     CO    49077    45314
4       DC     AK    11945     2729
5       DE     DC    14602    11945
6      FSM     AK        0     2729
7      FSM     AS        0        0
8       GA     AZ   175052   165934
9       GU     AK      351     2729
10      GU     AS      351        0
11      GU    FSM      351        0
12      HI     AK     1647     2729
13      HI     AS     1647        0
14      HI    FSM     1647        0
15      HI     GU     1647      351
16      IA     AR    42782    40181
17      IA     CO    42782    45314
18      IA     CT    42782    49077
19      ID     DC    19222    11945
20      ID     DE    19222    14602
21      IL     AZ   174973   165934
22      IL     GA   174973   175052
23      KS     ID    26172    19222
24      KY     ID    28126    19222
25      KY     KS    28126    26172
26      MA     LA   116182   111038
27      MD     AL    86285    82530
28      ME     AK     3866     2729
29      ME     AS     3866        0
..     ...    ...      ...      ...
236     WV     GU     6173      351
237     WV     HI     6173     1647
238     WV     ME     6173     3866
239     WV     MP     6173       40
240     WV     MT     6173     3667
241     WV     ND     6173     6227
242     WV     NH     6173     6500
243     WV     PR     6173    16061
244     WV     PW     6173        0
245     WV    RMI     6173        0
246     WV     SD     6173     8492
247     WV     VI     6173      385
248     WV     VT     6173     1405
249     WY     AK     2589     2729
250     WY     AS     2589        0
251     WY     DC     2589    11945
252     WY    FSM     2589        0
253     WY     GU     2589      351
254     WY     HI     2589     1647
255     WY     ME     2589     3866
256     WY     MP     2589       40
257     WY     MT     2589     3667
258     WY     ND     2589     6227
259     WY     NH     2589     6500
260     WY     PW     2589        0
261     WY    RMI     2589        0
262     WY     SD     2589     8492
263     WY     VI     2589      385
264     WY     VT     2589     1405
265     WY     WV     2589     6173

[266 rows x 4 columns]

#### Using a Cross Product

This question came up during the most recent SQL workshop - what happens when you don't specify a JOIN condition. 
Are there times when you'd want the entire cross-product of two tables?

Yes, there are times when you want to take the entire cross product of two tables, to get every possible pairing of rows from two (or more) different tables. Keep in mind that this can produce extremely large result sets quickly! If you're dealing with large tables (or multiple tables), you may want to reconsider your query and try to find a way reduce the result set. 

We can either use a JOIN with no ON condition or use a CROSS JOIN. I prefer to use the CROSS JOIN syntax, as it makes it clear in the query that I really do intend to take the full cross product, and haven't just forgotten to include an ON condition in my SQL query. Forgetting to include an ON clause in a JOIN is a fairly common SQL bug (actually, when my queries are taking an unexpectedly long time to run, this is the first thing I check). 

To illustrate, let's write a SQL query to find each state and state with the closest number of cases.

We'll start by taking the cross product of all states, and print out the difference between the case count for the two states

In [34]:
all_states_diff = pysqldf("""
SELECT
    a.abbr AS abbr_a,
    b.abbr AS abbr_b,
    ABS(a.total_cases - b.total_cases) diff
FROM
    df a
CROSS JOIN
    df b
WHERE
   abbr_a <> abbr_b
AND
    a.total_cases <> 0
AND 
    b.total_cases <> 0
""")

In [35]:
all_states_diff

abbr_a abbr_b     diff
0        AK     AL    79801
1        AK     AR    37452
2        AK     AZ   163205
3        AK     CA   463821
4        AK     CO    42585
5        AK     CT    46348
6        AK     DC     9216
7        AK     DE    11873
8        AK     FL   434138
9        AK     GA   172323
10       AK     GU     2378
11       AK     HI     1082
12       AK     IA    40053
13       AK     ID    16493
14       AK     IL   172244
15       AK     IN    60949
16       AK     KS    23443
17       AK     KY    25397
18       AK     LA   108309
19       AK     MA   113453
20       AK     MD    83556
21       AK     ME     1137
22       AK     MI    85229
23       AK     MN    50218
24       AK     MO    42094
25       AK     MP     2689
26       AK     MS    53075
27       AK     MT      938
28       AK     NC   113358
29       AK     ND     3498
...     ...    ...      ...
3162     US     MP  4339957
3163     US     MS  4284193
3164     US     MT  4336330
3165     US     NC  4223910
3166     US     ND  4333770
3167     US     NE  4314840
3168     US     NH  4333497
3169     US     NJ  4159702
3170     US     NM  4320206
3171     US     NV  4294846
3172     US     NY  4151368
3173     US    NYC  4114147
3174     US     OH  4253500
3175     US     OK  4305007
3176     US     OR  4322581
3177     US     PA  4229779
3178     US     PR  4323936
3179     US     RI  4321272
3180     US     SC  4255888
3181     US     SD  4331505
3182     US     TN  4240953
3183     US     TX  3945732
3184     US     UT  4300660
3185     US     VA  4252004
3186     US     VI  4339612
3187     US     VT  4338592
3188     US     WA  4285792
3189     US     WI  4285933
3190     US     WV  4333824
3191     US     WY  4337408

[3192 rows x 3 columns]

Next, we can write a query to find the state with the smallest difference for each state

There are a few ways to do this. One approach is to use MIN and GROUP BY. However, keep in mind the query below is allowable only in a few versions of SQLite. Generally, all terms in an aggregation query must be either in the aggregation function or in the GROUP BY clause. In the query below, abbr_b is in neither. 

This syntax is permitted and will work only with MIN and MAX aggregations, largely because MIN and MAX can be matched with a unique row, whereas SUM or AVG cannot. 

For more info, see 

In [36]:
pysqldf("""
SELECT 
    abbr_a, abbr_b, MIN(diff)
FROM
    all_states_diff
GROUP BY 
    abbr_a
""")

abbr_a abbr_b  MIN(diff)
0      AK     WY        140
1      AL     SC       1579
2      AR     UT        844
3      AZ     IL       9039
4      CA     FL      29683
5      CO     NV        163
6      CT     CO       3763
7      DC     DE       2657
8      DE     PR       1459
9      FL     CA      29683
10     GA     IL         79
11     GU     VI         34
12     HI     VT        242
13     IA     MO       2041
14     ID     RI        497
15     IL     GA         79
16     IN     MS       7874
17     KS     NE       1015
18     KY     KS       1954
19     LA     PA        820
20     MA     NC         95
21     MD     OH        212
22     ME     MT        199
23     MI     VA         35
24     MN     WI       1117
25     MO     NV        328
26     MP     GU        311
27     MS     WA       1599
28     MT     ME        199
29     NC     MA         95
30     ND     WV         54
31     NE     KS       1015
32     NH     ND        273
33     NJ     GA       5243
34     NM     ID        569
35     NV     CO        163
36     NY     NJ       8334
37    NYC     NY      37221
38     OH     MD        212
39     OK     UT       4347
40     OR     RI       1309
41     PA     LA        820
42     PR     OR       1355
43     RI     ID        497
44     SC     AL       1579
45     SD     NH       1992
46     TN     VA      11051
47     TX     FL      42602
48     US     CA    3873447
49     UT     AR        844
50     VA     MI         35
51     VI     GU         34
52     VT     HI        242
53     WA     WI        141
54     WI     WA        141
55     WV     ND         54
56     WY     AK        140

For a more generalizable query, you can use a LEFT JOIN. 

Notice that there is a difference between the this result set and the GROUP by approach above. This set will list rows with ties (ie., when the difference between a state and two or more other states is equal).

In [37]:
pysqldf("""
SELECT 
    a.abbr_a, a.abbr_b, a.diff
FROM 
    all_states_diff a                   
LEFT JOIN 
    all_states_diff b            
ON 
    A.abbr_b = b.abbr_b 
    AND 
    A.diff > b.diff
WHERE 
    b.diff is NULL 
""")

abbr_a abbr_b     diff
0      AK     WY      140
1      AL     SC     1579
2      AR     UT      844
3      CA     FL    29683
4      CA     US  3873447
5      CO     CT     3763
6      CO     NV      163
7      DE     DC     2657
8      FL     CA    29683
9      FL     TX    42602
10     GA     IL       79
11     GA     NJ     5243
12     GU     MP      311
13     GU     VI       34
14     HI     VT      242
15     ID     NM      569
16     ID     RI      497
17     IL     AZ     9039
18     IL     GA       79
19     KS     KY     1954
20     KS     NE     1015
21     LA     PA      820
22     MA     NC       95
23     MD     OH      212
24     ME     MT      199
25     MI     VA       35
26     MO     IA     2041
27     MS     IN     7874
28     MT     ME      199
29     NC     MA       95
30     ND     NH      273
31     ND     WV       54
32     NE     KS     1015
33     NH     SD     1992
34     NJ     NY     8334
35     NV     CO      163
36     NV     MO      328
37     NY    NYC    37221
38     OH     MD      212
39     OR     PR     1355
40     PA     LA      820
41     PR     DE     1459
42     RI     ID      497
43     RI     OR     1309
44     SC     AL     1579
45     UT     AR      844
46     UT     OK     4347
47     VA     MI       35
48     VA     TN    11051
49     VI     GU       34
50     VT     HI      242
51     WA     MS     1599
52     WA     WI      141
53     WI     MN     1117
54     WI     WA      141
55     WV     ND       54
56     WY     AK      140

### Exercise - Non-symmetric pairs

Some of these pairs repeat - for example, AK's closest match is WY at 140, and WY's closest match is AK. 

Which pairs are only work in one direction? For example, Texas's closest match is Florida, but Florida's closest match is California.
